src/handlers/upload.rs

```diff
@@ -391,35 +395,128 @@ pub fn install_forc_at_path(forc_version: &str, forc_path: &Path) -> Result<(),
         }
     };
 
+    for component in missing_components(forc_path, FORC_COMPONENTS) {
+        install_component(forc_version, forc_path, os, arch, component)?;
+
+        if !forc_path.join("bin").join(component.binary_name()).exists() {
+            return Err(UploadError::InvalidForcVersion(forc_version.to_string()));
```

Comment:
nit: a more appropriate error variant could be used here - i.e. `UploadError::ForcComponentBinaryMissing` for example.